tests/cupy_tests/lib_tests/test_polynomial.py

```diff
@@ -455,30 +462,100 @@ def test_polyroutine_diff_types_poly1d(self, xp, dtype1, dtype2):
 
 @testing.gpu
 @testing.parameterize(*testing.product({
-    'input': [[2, -1, -2], [-4, 10, 4]],
+    'type_l': ['poly1d', 'ndarray'],
+    'type_r': ['ndarray', 'numpy_scalar', 'python_scalar'],
 }))
-class TestRootsReal(unittest.TestCase):
+class TestPolyval(Poly1dTestBase):
 
-    @testing.for_signed_dtypes()
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-5)
+    def test_polyval(self, xp, dtype):
+        a1 = self._get_input(xp, self.type_l, dtype)
+        a2 = self._get_input(xp, self.type_r, dtype)
+        return xp.polyval(a1, a2)
+
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'type_l': ['numpy_scalar', 'python_scalar'],
+    'type_r': ['poly1d', 'ndarray', 'python_scalar', 'numpy_scalar'],
+}))
+class TestPolyvalInvalidTypes(Poly1dTestBase):
+
+    @testing.for_all_dtypes()
+    def test_polyval(self, dtype):
+        for xp in (numpy, cupy):
+            a1 = self._get_input(xp, self.type_l, dtype)
+            a2 = self._get_input(xp, self.type_r, dtype)
+            with pytest.raises(TypeError):
+                xp.polyval(a1, a2)
+
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'shape1': [(0,), (3,), (5,)],
+    'shape2': [(), (0,), (3,), (5,)]
+}))
+class TestPolyvalShapeCombination(unittest.TestCase):
+
+    @testing.for_all_dtypes()
     @testing.numpy_cupy_allclose(rtol=1e-6)
-    def test_roots_array(self, xp, dtype):
-        a = xp.array(self.input, dtype)
-        out = xp.roots(a)
-        return xp.sort(out)
+    def test_polyval(self, xp, dtype):
+        a = testing.shaped_arange(self.shape1, xp, dtype)
+        b = testing.shaped_arange(self.shape2, xp, dtype)
+        return xp.polyval(a, b)
 
-    @testing.for_signed_dtypes()
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'shape': [(), (0,), (3,), (5,)]
+}))
+class TestPolyvalInvalidShapeCombination(unittest.TestCase):
+
+    @testing.for_all_dtypes()
+    def test_polyval(self, dtype):
+        for xp in (numpy, cupy):
+            a = testing.shaped_arange((), xp, dtype)
+            b = testing.shaped_arange(self.shape, xp, dtype)
+            with pytest.raises(TypeError):
+                xp.polyval(a, b)
+
+
+@testing.gpu
+class TestPolyvalDiffTypes(unittest.TestCase):
+
+    @testing.for_all_dtypes_combination(names=['dtype1', 'dtype2'])
     @testing.numpy_cupy_allclose(rtol=1e-6)
-    def test_roots_poly1d(self, xp, dtype):
-        a = xp.array(self.input, dtype)
-        out = xp.roots(xp.poly1d(a))
-        return xp.sort(out)
+    def test_polyval_diff_types_array(self, xp, dtype1, dtype2):
```

Comment:
Please fix it.
```
>>> numpy.polyval(numpy.ones((2,), dtype='D'), numpy.ones((2,), dtype='F')).dtype
dtype('complex64')
>>> cupy.polyval(cupy.ones((2,), dtype='D'), cupy.ones((2,), dtype='F')).dtype
dtype('complex128')
```

Reply:
How? the tests pass locally

Reply:
`testing.for_all_dtypes_combination` does not check all combinations because of the following background.
https://docs.cupy.dev/en/stable/reference/generated/cupy.testing.for_dtypes_combination.html